sphinx/tutorial/cylc/furthertopics/broadcast.rst

```diff
@@ -0,0 +1,134 @@
+Broadcast
+=========
+
+This tutorial walks you through using ``cylc broadcast`` which can be used
+to change :ref:`task runtime configuration <tutorial-runtime>` in a
+running suite, on-the-fly.
+
+
+Purpose
+-------
+
+``cylc broadcast`` can be used to change any ``[runtime]`` setting whilst the
+suite is running.
+
+The standard use of ``cylc broadcast`` is to update the suite to an
+unexpected change in configuration, for example modifying the host a task
+runs on.
+
+
+Standalone Example
+------------------
+
+Create a new suite in the ``cylc-run`` directory called
+``tutorial-broadcast``::
+
+   mkdir ~/cylc-run/tutorial-broadcast
+   cd ~/cylc-run/tutorial-broadcast
+
+Copy the following configuration into a ``suite.rc`` file:
+
+.. code-block:: cylc
+
+   [scheduling]
+       initial cycle point = 1012
+       [[dependencies]]
+           [[[R1]]]
+               graph = wipe_log => announce
+           [[[PT1H]]]
+               graph = announce[-PT1H] => announce
+
+   [runtime]
+       [[wipe_log]]
+           # Delete any files in the suite's "share" directory.
+           script = rm "${CYLC_SUITE_SHARE_DIR}/*"
+
+       [[announce]]
+           script = echo "${CYLC_TASK_CYCLE_POINT} - ${MESSAGE}" >> "${FILE}"
+           [[[environment]]]
+               WORD = ni
+               MESSAGE = We are the knights who say \"${WORD}\"!
+               FILE = "${CYLC_SUITE_SHARE_DIR}/knights"
+
+We now have a suite with an ``announce`` task which runs every hour, writing a
+message to a log file (``share/knights``) when it does so. For the first cycle
+the log entry will look like this::
+
+   10120101T0000Z - We are the knights who say "ni"!
+
+The ``cylc broadcast`` task enables us to change runtime configuration whilst
+the suite is running. For instance we could change the value of the ``WORD``
+environment variable using the command::
+
+   cylc broadcast tutorial-broadcast -n announce -s "[environment]WORD=it"
+
+* ``tutorial-broadcast`` is the name of the suite.
+* ``-n announce`` tells cylc we want to change the runtime configuration of the
+  ``announce`` task.
+* ``-s "[environment]WORD=it"`` changes the value of the ``WORD`` environment
+  variable to ``it``.
+
+Run the suite then try using the ``cylc broadcast`` command to change the
+message::
+
+   cylc run tutorial-broadcast
+   cylc broadcast tutorial-broadcast -n announce -s "[environment]WORD=it"
+
+Inspect the ``share/knights`` file, you should see the message change at
```

Comment:
IMPORTANT: I follow the commands given and find I have no 'share/knights' file - the share directory is empty! I tried running a new suite with the same 'suite.rc' file but the script for 'wipe_log' commented out and this gives the correct output, so it is something to do with the wipe_log task running once at the start deleting the contents of the 'share' directory - possibly ``FILE = "${CYLC_SUITE_SHARE_DIR}/knights"`` doesn't do anything because you need to make the knights directory before writing to it? Not sure what it could be otherwise as wipe_log only runs at the start so shouldn't have an influence after.

Reply:
Okay, have now investigated this and it works as it should (i.e. writing to the 'knights' file in the share directory) with the script for the 'announce' task changed to ``mkdir "${FILE}" && echo ...``.

Reply:
Adding a ``touch "${FILE}"`` command into the 'wipe_log' script doesn't seem to resolve this - trying to investigate presently.

Reply:
Just discovered '>>' creates a file should it not exist so puzzled as to why the 'share/knights' file isn't there!

Reply:
I'll fix this in my next push, probs best to do `rm file || true`.

Reply:
Not quite sure what's happening there, but changing the 'wipe_log' task script to ``ls "${CYLC_SUITE_SHARE_DIR}/*" | grep -v knights | xargs rm; > knights`` to empty the 'knights' file instead of removing it along with the rest of contents of the 'share' sub-directory works.

Reply:
Oops your comment and commit didn't appear before I commented again, sounds like you know what's happening so nevermind!

Reply:
Yes that works fine 👍

Reply:
Actually I have just noticed your fix, while not removing the knights file to give the initial issue I brought up, does mean that the log doesn't seem to get wiped every-time a suite is stopped and rerun, which is the whole point of the 'wipe_log' task I assume. So you might want to change that task script again to something different.